formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Decoder.kt

```diff
@@ -0,0 +1,636 @@
+        }
+        readByte()
+        return ans
+    }
+
+    fun startArray(tag: ULong) = startArray(ulongArrayOf(tag))
+
+    fun startArray(tags: ULongArray? = null) = startSized(tags, BEGIN_ARRAY, HEADER_ARRAY, "array")
+
+    fun startMap(tag: ULong) = startMap(ulongArrayOf(tag))
+
+    fun startMap(tags: ULongArray? = null) = startSized(tags, BEGIN_MAP, HEADER_MAP, "map")
+
+    private fun startSized(
+        tags: ULongArray?,
+        unboundedHeader: Int,
+        boundedHeaderMask: Int,
+        collectionType: String
+    ): Int {
+        processTags(tags)
+        if (curByte == unboundedHeader) {
+            skipByte(unboundedHeader)
+            return -1
+        }
+        if ((curByte and 0b111_00000) != boundedHeaderMask)
+            throw CborDecodingException("start of $collectionType", curByte)
+        val size = readNumber().toInt()
+        readByte()
+        return size
+    }
+
+    fun isEnd() = curByte == BREAK
+
+    fun end() = skipByte(BREAK)
+
+    fun nextByteString(tags: ULongArray? = null): ByteArray {
+        processTags(tags)
+        if ((curByte and 0b111_00000) != HEADER_BYTE_STRING)
+            throw CborDecodingException("start of byte string", curByte)
+        val arr = readBytes()
+        readByte()
+        return arr
+    }
+
+    fun nextString(tag: ULong) = nextString(ulongArrayOf(tag))
+    fun nextString(tags: ULongArray? = null) = nextTaggedString(tags).first
+
+    //used for reading the tag names and names of tagged keys (of maps, and serialized classes)
+    private fun nextTaggedString(tags: ULongArray?): Pair<String, ULongArray?> {
+        val collectedTags = processTags(tags)
+        if ((curByte and 0b111_00000) != HEADER_STRING)
+            throw CborDecodingException("start of string", curByte)
+        val arr = readBytes()
+        val ans = arr.decodeToString()
+        readByte()
+        return ans to collectedTags
+    }
+
+    private fun readBytes(): ByteArray =
+        if (curByte and 0b000_11111 == ADDITIONAL_INFORMATION_INDEFINITE_LENGTH) {
+            readByte()
+            readIndefiniteLengthBytes()
+        } else {
+            val strLen = readNumber().toInt()
+            input.readExactNBytes(strLen)
+        }
+
+    @SuppressAnimalSniffer
+    private fun processTags(tags: ULongArray?): ULongArray? {
+        var index = 0
+        val collectedTags = mutableListOf<ULong>()
+        while ((curByte and 0b111_00000) == HEADER_TAG) {
+            val readTag = readNumber().toULong() // This is the tag number
+            collectedTags += readTag
+            // value tags and object tags are intermingled (keyTags are always separate)
+            // so this check only holds if we verify both
+            if (verifyObjectTags) {
+                tags?.let {
+                    if (index++ > it.size) throw CborDecodingException("More tags found than the ${it.size} tags specified")
```

Comment:
Shouldn't it be `index++ >= it.size` since there is no element at `it[it.size]`? I suggest you write a test for a common Cbor validation errors

Reply:
Wow! Awesome catch!  I really though I had all mistagged cases covered. that one is fixed and has a test case for it too in 6cd632416cabe12e15c6094e6e33cf3b78b9d262.